File: frontend/src/api/themes.ts
export interface Theme {
  id: number;
  name: string;
  etf_symbol: string;
  description: string | null;
  stock_count: number;
}

export interface ThemeAssignment {
  theme_id: number;
  theme_name: string;
  source: "manual" | "manual_ai" | "automatic_ai";
  reasoning: string | null;
  model: string | null;
  assigned_at: string;
}

export interface ThemeTicker {
  symbol: string;
  name: string | null;
  description: string | null;
  assignments: ThemeAssignment[];
}

export interface ThemeSuggestion {
  symbol: string;
  themes: string[];
  reasoning?: string | null;
}

export interface AiCapability {
  enabled: boolean;
  model: string | null;
  batch_size: number | null;
}

export interface ThemeAiJob {
  id: number;
  status: "pending" | "running" | "completed" | "failed" | "applied";
  symbols: string[];
  model: string;
  prompt: string;
  response: string | null;
  suggestions: ThemeSuggestion[] | null;
  error: string | null;
  created_at: string;
  updated_at: string;
}

export interface ThemeAiJobSummary {
  id: number;
  status: ThemeAiJob["status"];
  symbol_count: number;
  model: string;
  updated_at: string;
}

interface ThemeInput {
  name: string;
  etf_symbol: string;
  description: string | null;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(path, init);
  if (!response.ok) {
    const body = (await response.json().catch(() => undefined)) as { error?: string } | undefined;
    throw new Error(body?.error ?? `Theme request failed: HTTP ${response.status}`);
  }
  return response.json() as Promise<T>;
}

const json = (body: unknown): RequestInit => ({
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify(body),
});

export const fetchThemes = () => request<Theme[]>("/api/themes");
export const fetchThemeTickers = () => request<ThemeTicker[]>("/api/theme-tickers");
export const fetchThemeTicker = (symbol: string) =>
  request<ThemeTicker>(`/api/theme-tickers/${encodeURIComponent(symbol)}`);
export const fetchAiCapability = () => request<AiCapability>("/api/theme-ai/capability");

export const addThemeTicker = (symbol: string) =>
  request<{ ok: boolean }>("/api/theme-tickers", {
    method: "POST",
    ...json({ symbol }),
  });

export const createTheme = (input: ThemeInput) =>
  request<{ id: number }>("/api/themes", { method: "POST", ...json(input) });

export const updateTheme = (id: number, input: ThemeInput) =>
  request<{ ok: boolean }>(`/api/themes/${id}`, { method: "PUT", ...json(input) });

export const deleteTheme = (id: number) =>
  request<{ ok: boolean }>(`/api/themes/${id}`, { method: "DELETE" });

export const replaceTickerThemes = (symbol: string, themeIds: number[]) =>
  request<{ ok: boolean }>(`/api/theme-tickers/${encodeURIComponent(symbol)}`, {
    method: "PUT",
    ...json({ theme_ids: themeIds }),
  });

export const generateThemePrompt = (symbols: string[]) =>
  request<{ prompt: string }>("/api/theme-ai/prompt", {
    method: "POST",
    ...json({ symbols }),
  });

export const parseThemeSuggestions = (response: string) =>
  request<ThemeSuggestion[]>("/api/theme-ai/parse", {
    method: "POST",
    ...json({ response }),
  });

export const suggestThemeAssignments = (symbols: string[]) =>
  request<ThemeSuggestion[]>("/api/theme-ai/suggest", {
    method: "POST",
    ...json({ symbols }),
  });

export const createAutomaticJobs = (symbols: string[]) =>
  request<{ ids: number[] }>("/api/theme-ai/jobs", {
    method: "POST",
    ...json({ symbols }),
  });

export const fetchThemeAiJobs = () => request<ThemeAiJobSummary[]>("/api/theme-ai/jobs");

export const fetchThemeAiJob = (id: number) => request<ThemeAiJob>(`/api/theme-ai/jobs/${id}`);

export const applyThemeAiJob = (id: number) =>
  request<{ ok: boolean }>(`/api/theme-ai/jobs/${id}/apply`, { method: "POST" });

export const deleteThemeAiJob = (id: number) =>
  request<{ ok: boolean }>(`/api/theme-ai/jobs/${id}`, { method: "DELETE" });

export const applyThemeSuggestions = (
  suggestions: ThemeSuggestion[],
  source: "manual_ai" | "automatic_ai",
) =>
  request<{ ok: boolean }>("/api/theme-ai/apply", {
    method: "POST",
    ...json({ suggestions, source }),
  });

File: src/models/themes.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize)]
pub struct Theme {
    pub id: i64,
    pub name: String,
    pub etf_symbol: String,
    pub description: Option<String>,
    pub stock_count: i64,
}

#[derive(Clone, Debug, Serialize)]
pub struct ThemeTicker {
    pub symbol: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub industries: Vec<ThemeTickerIndustry>,
    pub assignments: Vec<ThemeAssignment>,
    pub automatic_processed: bool,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct ThemeTickerIndustry {
    pub key: String,
    pub name: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ThemeSuggestion {
    pub symbol: String,
    pub themes: Vec<String>,
    #[serde(default)]
    pub reasoning: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ThemeSuggestionError {
    pub symbol: Option<String>,
    pub error: String,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ThemeAiJobStatus {
    Pending,
    Running,
    Completed,
    PartiallyFailed,
    Failed,
    Applied,
}

#[derive(Clone, Debug, Serialize)]
pub struct ThemeAiJob {
    pub id: i64,
    pub status: ThemeAiJobStatus,
    pub symbols: Vec<String>,
    pub model: String,
    pub prompt: String,
    pub response: Option<String>,
    pub suggestions: Option<Vec<ThemeSuggestion>>,
    pub validation_errors: Vec<ThemeSuggestionError>,
    pub error: Option<String>,
    pub retry_of_job_id: Option<i64>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Clone, Debug, Serialize)]
pub struct ThemeAiJobSummary {
    pub id: i64,
    pub status: ThemeAiJobStatus,
    pub symbol_count: i64,
    pub model: String,
    pub updated_at: DateTime<Utc>,
}

#[derive(Clone, Debug, Serialize)]
pub struct ThemeAssignment {
    pub theme_id: i64,
    pub theme_name: String,
    pub source: AssignmentSource,
    pub reasoning: Option<String>,
    pub model: Option<String>,
    pub assigned_at: DateTime<Utc>,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AssignmentSource {
    Manual,
    ManualAi,
    AutomaticAi,
}

impl AssignmentSource {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Manual => "manual",
            Self::ManualAi => "manual_ai",
            Self::AutomaticAi => "automatic_ai",
        }
    }
}
